Automattic/wp-calypso
# Welcome Tour NEXT: Apply step repositioning

#### Changes proposed in this Pull Request

This PR sets the ground for moving/positioning the tour near the element the various steps refer to - so going from a static tour to a more step-by-step guided tour. Three main thngs being introduced:

1. [popper.js](https://popper.js.org) to move the tour near reference elements on the page.
2. A dark overlay that blocks mouse events from reaching the editor.
3. Focus trapping e.g. navigating via Tab or Shift+Tab keys traps navigation inside the tour, so keeps it in a loop inside the tour when the tour is maximised.

![Kapture 2021-10-12 at 17 35 40](https://user-images.githubusercontent.com/1705499/136976971-606d7156-6376-4277-a74c-877d827cfcaf.gif)

#### Testing instructions

- Sandbox a site, checkout branch, and yarn dev --sync from apps/editing-toolkit to sync sandbox.
- Go to the editor, open the sidebar, and from there click on "Welcome Guide".
- Tour should be focused and keyboard navigation enabled.
- Focus should be on the "Try it out!" button.
- Stepping through the steps should move them closer to the element they refer to (as in the [Figma mocks](Mh2ZK4ZkifXClvLEG5imNc-fi-596%3A15786)).
- Navigating via Tab key should trap the focus within the tour if the tour is expanded (not minified).

TODO

- [x] Focus trapping is currently incomplete/broken. It works on the first step as expected, but moving to the next step, it stops functioning. 
- [x] Might be best to always focus the "Next" button on every new card/step render (which may actually fix the above too).
- [x] Follow up (create issue): Update the reference elements between mobile & desktop.
- [x] Follow up (create issue): Consider using [use-constrained-tabbing](https://github.com/WordPress/gutenberg/tree/trunk/packages/compose/src/hooks/use-constrained-tabbing) hook from core. 
- [x] Follow up (create issue): E2E test.
- [x] Follow up (create issue): spotlight effect.






Related to #56353
